linked_list/linked_list.py

```diff
@@ -86,11 +165,26 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+
+        if not self.head:
+            return None
+
+        previous = None
+        current = self.head
+        next_node = None
+
+        # make the list doubly linked by adding real values to .previous
+        while current:

```

Comment:
previous here is just a placeholder variable here so that the next iteration through the while loop, you know what the previous node to the current node you are iterating over will be. 

That's why we set `previous = current`. Because the next iteration through the while loop, the previous node will be the former current node.